ShaderGenerator/ShaderConfiguration.cpp

```diff
@@ -230,29 +237,32 @@ namespace ShaderGenerator
         value.Key |= indices[i] << offset;
         offset += option->KeyLength();
       }
-      results.push_back(value);
+      results.push_back(move(value));
 
       //Iterate
       indices[currentIndex]++;
       auto roll = false;
       while (!done && indices[currentIndex] == options[currentIndex]->ValueCount())
       {
         roll = true;
-        currentIndex++;
-        if (currentIndex == indices.size())
+        if (currentIndex == 0)
         {
           done = true;
           continue;
         }
+        else
+        {
+          currentIndex--;
+        }
 
-        for (size_t i = 0; i < currentIndex; i++)
+        for (size_t i = currentIndex+1; i < indices.size(); i++)
         {
           indices[i] = 0;
         }
         indices[currentIndex]++;
       }
 
-      if (roll) currentIndex = 0;
+      if (roll) currentIndex = indices.size() - 1;
```

Comment:
The goal was to put those option permutations  that share the first few options values near each other in the output. I could achieve this by reversing the indices here. Reversing the option vector would have worked, but it would have affected other parts for example the header generation. I could have reversed twice inside the function but I did not find that elegant.